File: src/CmdCaligraph.cpp
// CmdCaligraph.cpp - This file is part of QPlot

/* QPlot - Publication quality 2D graphs with dual coordinate systems
   Copyright (C) 2014  Daniel Wagenaar
  
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
  
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

// CmdCaligraph.C

#include "CmdCaligraph.H"
#include "Rotate.H"

#include <QDebug>
#include <math.h>

static CBuilder<CmdCaligraph> cbCaligraph("caligraph");

bool CmdCaligraph::usage() {
  return error("Usage: caligraph xdata ydata widthdata");
}

bool CmdCaligraph::parse(Statement const &s) {
  int id1 = s.nextIndex(1);
  int id2 = s.nextIndex(id1);
  int id3 = s.nextIndex(id2);
  if (id3==s.length()
      && s.isNumeric(1) && s.isNumeric(id1) && s.isNumeric(3)
      && s.data(1).size()==s.data(id1).size() 
      && s.data(1).size()==s.data(id2).size() )
    return true;
  else
    return usage();
}

QRectF CmdCaligraph::dataRange(Statement const &s) {
  QVector<double> const &xdata = s.data(1);
  QVector<double> const &ydata = s.data(s.nextIndex(1));

  if (xdata.isEmpty())
    return QRectF();

  double minx = xdata[0];
  double maxx = xdata[0];
  double miny = ydata[0];
  double maxy = ydata[0];

  for (int k=1; k<xdata.size(); k++)
    if (xdata[k]<minx)
      minx=xdata[k];
    else if (xdata[k]>maxx)
      maxx=xdata[k];
  for (int k=1; k<ydata.size(); k++)
    if (ydata[k]<miny)
      miny=ydata[k];
    else if (ydata[k]>maxy)
      maxy=ydata[k];
  return QRectF(QPointF(minx,miny), QPointF(maxx,maxy));
}

void CmdCaligraph::render(Statement const &s, Figure &f, bool dryrun) {
  int id1 = s.nextIndex(1);
  int id2 = s.nextIndex(id1);
  QVector<double> const &xdata = s.data(1);
  QVector<double> const &ydata = s.data(id1);
  QVector<double> const &wdata = s.data(id2);
  if (xdata.isEmpty()) {
    f.setBBox(QRectF());
    return;
  }

  /* Each point on the line is replaced by three: two on the outer corner
     and one on the inner, except for the first and last points, which
     get replaced by only two each. */
  int N = xdata.size();
  int L = 3*N-2;
  int l1 = 0;
  int l2 = L;
  QPolygonF poly(L);

  QPointF pprev, pthis, pnext;
  QPointF dprev, dnext;
  double thetaprev = 0; // this value not used unless N=1
  double thetanext = 0; // this value not used unless N=1
  for (int k=0; k<N; k++) {
    if (k>0) 
      pprev = pthis;
    if (k>0)
      pthis = pnext;
    else 
      pthis = f.map(xdata[k],ydata[k]);
    if (k<N-1)
      pnext = f.map(xdata[k+1],ydata[k+1]);

    if (k>0) 
      dprev = dnext;
    if (k<N-1)
      dnext = pnext - pthis;

    if (k>0)
      thetaprev = thetanext;
    if (k<N-1)
      thetanext = atan2(dnext.y(), dnext.x());

    double w = pt2iu(wdata[k]/2);
    if (k==0) {
      QPointF ddnext = w * QPointF(-sin(thetanext), cos(thetanext));
      poly[l1++] = pthis+ddnext;
      poly[--l2] = pthis-ddnext;
    } else if (k==N-1) {
      QPointF ddprev = w * QPointF(-sin(thetaprev), cos(thetaprev));
      poly[l1++] = pthis+ddprev;
      poly[--l2] = pthis-ddprev;
    } else {
      QPointF ddprev = w * QPointF(-sin(thetaprev), cos(thetaprev));
      QPointF ddnext = w * QPointF(-sin(thetanext), cos(thetanext));
      if (dnext.y()*dprev.x() > dnext.x()*dprev.y()) {
	poly[l1++] = pthis+ddprev/2+ddnext/2;
	poly[--l2] = pthis-ddprev;
	poly[--l2] = pthis-ddnext;
      } else {
	poly[l1++] = pthis+ddprev;
	poly[l1++] = pthis+ddnext;
	poly[--l2] = pthis-ddprev/2-ddnext/2;
      }
    }
  }
  if (l1!=l2)
    qDebug() << "Caligraph: point miscount";

  QRectF bbox = poly.boundingRect();
  f.setBBox(bbox);

  if (dryrun)
    return;

  QPen pen = f.painter().pen();
  QBrush brush = f.painter().brush();
  f.painter().setPen(Qt::NoPen);
  f.painter().setBrush(pen.color());
  f.painter().drawPolygon(poly);
  f.painter().setPen(pen);
  f.painter().setBrush(brush);
}

File: python/qplot_backend/cmdcaligraph.py
# cmdcaligraph.py - This file is part of QPlot

# QPlot - Publication quality 2D graphs with dual coordinate systems

from __future__ import annotations

import math
from typing import TYPE_CHECKING

from PyQt6.QtCore import Qt, QPointF, QRectF
from PyQt6.QtGui import QPolygonF, QBrush

from .command import Command
from .token_ import Token
from .factor import pt2iu

if TYPE_CHECKING:
    from .statement import Statement
    from .figure import Figure


@Command.register("caligraph")
class CmdCaligraph(Command):
    """Draw a calligraphic line of variable width.

    Syntax:
        caligraph xdata ydata widthdata
    """

    def _usage(self) -> bool:
        return self.error("Usage: caligraph xdata ydata widthdata")

    def parse(self, s: Statement) -> bool:
        id1 = s.next_index(1)
        id2 = s.next_index(id1)
        id3 = s.next_index(id2)
        if (id3 == len(s)
                and s.is_numeric(1) and s.is_numeric(id1) and s.is_numeric(id2)
                and len(s.data(1)) == len(s.data(id1))
                and len(s.data(1)) == len(s.data(id2))):
            return True
        return self._usage()

    def data_range(self, s: Statement) -> QRectF:
        xdata = s.data(1)
        ydata = s.data(s.next_index(1))
        if not xdata:
            return QRectF()
        return QRectF(
            QPointF(min(xdata), min(ydata)),
            QPointF(max(xdata), max(ydata)),
        )

    def render(self, s: Statement, f: Figure, dryrun: bool) -> None:
        id1 = s.next_index(1)
        id2 = s.next_index(id1)
        xdata = s.data(1)
        ydata = s.data(id1)
        wdata = s.data(id2)

        if not xdata:
            f.set_bbox(QRectF())
            return

        # Each interior point expands to three polygon vertices:
        # two on the outer corner, one on the inner.
        # First and last points expand to two each.
        N = len(xdata)
        L = 3 * N - 2
        poly = [QPointF(0, 0)] * L  # pre-allocate, filled from both ends
        l1 = 0                       # front pointer (outer edge)
        l2 = L                       # back pointer  (inner edge)

        pprev = QPointF()
        pthis = QPointF()
        pnext = QPointF()
        dprev = QPointF()
        dnext = QPointF()
        theta_prev = 0.0
        theta_next = 0.0

        for k in range(N):
            if k > 0:
                pprev = pthis
            if k > 0:
                pthis = pnext
            else:
                pthis = f.map(xdata[k], ydata[k])
            if k < N - 1:
                pnext = f.map(xdata[k + 1], ydata[k + 1])

            if k > 0:
                dprev = dnext
            if k < N - 1:
                dnext = pnext - pthis

            if k > 0:
                theta_prev = theta_next
            if k < N - 1:
                theta_next = math.atan2(dnext.y(), dnext.x())

            w = pt2iu(wdata[k] / 2)

            if k == 0:
                dd = w * QPointF(-math.sin(theta_next), math.cos(theta_next))
                poly[l1] = pthis + dd;  l1 += 1
                l2 -= 1;                poly[l2] = pthis - dd

            elif k == N - 1:
                dd = w * QPointF(-math.sin(theta_prev), math.cos(theta_prev))
                poly[l1] = pthis + dd;  l1 += 1
                l2 -= 1;                poly[l2] = pthis - dd

            else:
                dd_prev = w * QPointF(-math.sin(theta_prev), math.cos(theta_prev))
                dd_next = w * QPointF(-math.sin(theta_next), math.cos(theta_next))

                # Cross product sign determines which side is convex
                if dnext.y() * dprev.x() > dnext.x() * dprev.y():
                    poly[l1] = pthis + dd_prev / 2 + dd_next / 2;  l1 += 1
                    l2 -= 1;  poly[l2] = pthis - dd_prev
                    l2 -= 1;  poly[l2] = pthis - dd_next
                else:
                    poly[l1] = pthis + dd_prev;  l1 += 1
                    poly[l1] = pthis + dd_next;  l1 += 1
                    l2 -= 1;  poly[l2] = pthis - dd_prev / 2 - dd_next / 2

        if l1 != l2:
            import sys
            print("Caligraph: point miscount", file=sys.stderr)

        qpoly = QPolygonF(poly)
        bbox = qpoly.boundingRect()
        f.set_bbox(bbox)

        if dryrun:
            return

        painter = f.painter()
        pen = painter.pen()
        brush = painter.brush()
        painter.setPen(Qt.PenStyle.NoPen)
        painter.setBrush(QBrush(pen.color()))
        painter.drawPolygon(qpoly)
        painter.setPen(pen)
        painter.setBrush(brush)
